rfcs/90-re-support-initialization-via-connection-string.md

```diff
@@ -0,0 +1,152 @@
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from connection string (We rename config string to connection string since this name is more common):
+
+We add the following APIs:
+```go
+func NewServicerFromString(connStr string) (types.Servicer, error) {}
+func NewStoragerFromString(connStr string) (types.Storager, error) {}
+```
+
+### Format
+
+The format of the connection string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+- At least one of name and work_dir MUST be present.
+- For the `key=value` pairs, if `=value` is missing, we just ignore the pair. But `key=` means a pair with a blank value.
+
+So a valid connection string could be:
+
+- `s3://bucket_name`
+- `s3://bucket_name/prefix`
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&endpoint=http://bucket_name.s3-website-Region.amazonaws.com`
+- `fs:///tmp`
+
+#### Parseable Value Types
+
+- `string`
+- `bool`
+- `int`
+- `int64`
+- `[]byte`: input should be base-64 encoded
+
+For types like `credential` and `endpoint`, they are parsed in specific service's `new` functions, and treated as plain string here.
+
+
+In the future, we may consider supporting complex types like `struct` (`DefaultStoragePairs`).
+
+#### Escaping?
+
+We don't require (and don't allow) users to escape the options. However, `&` SHOULD NOT be used since we use it as the delimiter. Other characters are OK (`/`, `=`, `+`, ...).
+
+In the future, we may consider escaping and enabling `&`.
+
+### Implementation
+
+`New*FromString` will first split connection string into `(ty string, m map[string]string)`, and then parse it into `ps []Pairs`, and finally call `New*(ty, ps)`.
+To support this, we have to know that a name is a pair (global or service), and its type, so we implement:
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+// RegisterServiceSchema will register a service's pair map.
+//
+// Users SHOULD NOT call this function.
+func RegisterServiceSchema(ty string, m map[string]string) {}
```

Comment:
So we only support init a `Storager` from the connection string? Do we need to support init `Servicer`? `RegisterServiceSchema` may be confused between `servicer` and `storager`?

> BTW, It's OK to support storage only in this GSP if we have left some space for service support.

Reply:
There's no difference between `servicer` and `storager`.

I proviced `NewServicerFromString` and `NewStoragerFromString`. They both call `parseString(connStr)` which relys on `servicePairMaps`, which is initialized by `RegisterServiceSchema`

Reply:
Well, I got it.

How about rename `services.RegisterServiceSchema` to `services.RegisterSchema` so that we can avoid this confusion?

Reply:
Something missing: the format could be different for service and storage.

`s3` service doesn't need `name` nor `work_dir`, the examples may be:

```
`s3://?credential=hmac:xxxx:xxxx&endpoint=http://s3.us-east-2.amazonaws.com`
```

Reply:
> the format could be different for service and storage.

Oops. Then we just allow both of them to be missing? Let this be checked in generated service code, like other `required` pairs.

Reply:
I think only `name` is required for `Storage`, if the name exists, it's a storage, or it's a service. And `work_dir` has a default value `/`.

For our formats, It's OK to add omit the workdir like:

- `s3://bucketname?credential=hmac:xxxx:xxxx&endpoint=http://s3.us-east-2.amazonaws.com`
- `fs://` & `fs:///`

Allow omit or not are both OK to me, I respect your decision.